src/backend/distributed/operations/shard_rebalancer.c

```diff
@@ -449,10 +449,12 @@ citus_shard_cost_by_disk_size(PG_FUNCTION_ARGS)
 	uint32 connectionFlag = 0;
 	PGresult *result = NULL;
 	bool raiseErrors = true;
+	bool optimizePartitionCalculations = true;
 	ShardInterval *shardInterval = LoadShardInterval(shardId);
-	List *colocatedShardList = ColocatedShardIntervalList(shardInterval);
+	List *colocatedShardList = ColocatedNonPartitionShardIntervalList(shardInterval);
```

Comment:
As the [flamegraph](https://github.com/citusdata/citus/issues/4740#issuecomment-803058591) indicates, ~2/5 of the speed up we get by this pr is because `ColocatedNonPartitionShardIntervalList`. Because with that, we simply get rid of most of the costly operations in the loop in `ColocatedShardIntervalList`. In this new function, we just skip with `continue` if it's a child table. So we don't get the cache entries, we don't copy shard intervals and we don't sort them. I know it's not very pretty, but I think it's worth.